### PR TITLE
Migrate HitPattern to use TrackerTopology

### PR DESCRIPTION
**Pull request by makortel**
This PR migrates HitPattern to use TrackerTopology instead of the tracker DetId classes. There were two kind of use-cases for filling HitPattern from DetIds:
1. DetId originates from a hit
2. DetId is constructed on the fly from another packed HitPattern-like format
   1. Old HitPattern (pre-#4455)
   2. PackedCandidate

Case 1 was straightforward to solve by percolating TrackerTopology to all necessary places to be passed to HitPattern.

Case 2 was more tricky, as it was not possible/practical to percolate TrackerTopology. Instead, I decided to skip the creation of a "dummy" DetId altogether, and added a new function to deliver the required information (subdet, layer, stereo bit) to HitPattern (`appendTrackerHit()` for tracker, corresponding `appendMuonHit()` still employs DetId to encode the information as that seemed to be the simplest solution for now).

Tested for 1 and 2.ii in CMSSW_7_5_X_2015-06-14-2300 + #9630 with wf 11325 (#9630/#9801 needed to verify 2.ii, i.e. CMSSW_7_6_X_2015-06-29-1100 or later). Case 2.i was tested separately with a file from /RelValTTbar_13/CMSSW_7_1_0-POSTLS171_V15-v1/GEN-SIM-RECO (predating #4455 so that the iorule is activated) by printing out all properties of HitPattern first in the IB, then in the IB+this branch, and finally making a diff of the two printouts (showing no differences).

@rovere @VinInn

### CODE REVIEW

**Comment by makortel** [comment]
FYI @boudoul @venturia too

**Comment by makortel** [comment]
This PR "conflicts" with #9917 such that whichever gets merged first, the other one needs to migrate `TrackFromSeedProducer` (added in #9917) to use the new version of `TrackBase::appendHits()` introduced here.

**Comment by slava77** [comment]
@cmsbuild please test

File: DataFormats/TrackReco/src/classes_def.xml [hunk 1]
```diff
@@ -107,26 +107,7 @@
...
-                        detId = TIDDetId(DONT_CARE, layer, DONT_CARE, DONT_CARE, DONT_CARE, stereo);
-                        break;
-                    case StripSubdetector::TOB:
-                        detId = TOBDetId(layer, DONT_CARE, DONT_CARE, DONT_CARE, stereo);
-                        break;
-                    case StripSubdetector::TEC:
-                        detId = TECDetId(DONT_CARE, layer, DONT_CARE, DONT_CARE, DONT_CARE, DONT_CARE, stereo);
-                        break;
-                    }
+                    return newObj->appendTrackerHit(subdet, layer, stereo, hitType);
```
Comment (slava77) [inline DataFormats/TrackReco/src/classes_def.xml]: nice reduction

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/9959	1c0883c
- code looks fine (I didn't try to parse the static analyzer results, since it reports everything else in the release instead of this change alone; recompiled as dependent)
- jenkins tests passed and comparisons with baseline show no difference
- the interference with #9917 is noted and I would suggest to merge this one first due to a larger code base touched here. ( #9806 will conflict as well)

**Comment by lveldere** [comment]
+1

**Comment by deguio** [comment]
+1

**Pull request merged**